examples/async/index.js

```diff
@@ -1,15 +1,43 @@
 import 'babel-polyfill'
 import React from 'react'
-import { render } from 'react-dom'
+import ReactDOM from 'react-dom'
 import { Provider } from 'react-redux'
-import App from './containers/App'
 import configureStore from './store/configureStore'
 
 const store = configureStore()
+const rootEl = document.getElementById('root')
 
-render(
-  <Provider store={store}>
-    <App />
-  </Provider>,
-  document.getElementById('root')
-)
+let render = () => {
+  const App = require('./containers/App').default
+  ReactDOM.render(
+    <Provider store={store}>
+      <App />
+    </Provider>,
+    rootEl
+  )
+}
+
+if (module.hot) {
+  // Support hot reloading of components
+  // and display an overlay for runtime errors
+  const renderApp = render
+  const renderError = (error) => {
+    const RedBox = require('redbox-react')
+    ReactDOM.render(
+      <RedBox error={error} />,
+      rootEl
+    )
+  }
+  render = () => {
+    try {
+      renderApp()
+    } catch (error) {
+      renderError(error)
+    }
+  }
+  module.hot.accept('./containers/App', () => {
+    setTimeout(render)
```

Comment:
Line 11 `const App = require('./containers/App').default` always gets the updated component. Inline require calls do that by their very nature of being a function instead of a top-level `import`. The fact that the module isn't passed into `render` doesn't matter in this case because as soon as the render-function is run, the variable definition inside `render` in turn runs the `require`-function that triggers a re-require of the module.

Even though the code inside `render` is the same, that doesn't mean that the functions that are called within render don't produce a different result. Perhaps you're confusing class-constructors and regular functions.